api/src/exam-environment/routes/exam-environment.ts

```diff
@@ -0,0 +1,578 @@
+
+/**
+ * Wrapper for endpoints related to the exam environment desktop app.
+ *
+ * Requires exam environment authorization token to be validated.
+ */
+export const examEnvironmentValidatedTokenRoutes: FastifyPluginCallbackTypebox =
+  (fastify, _options, done) => {
+    // TODO: Is there any reason to delete the token without generating a new one?
+    fastify.post(
+      '/exam-environment/exam/generated-exam',
+      {
+        schema: schemas.examEnvironmentPostExamGeneratedExam
+      },
+      postExamGeneratedExamHandler
+    );
+    fastify.post(
+      '/exam-environment/exam/attempt',
+      {
+        schema: schemas.examEnvironmentPostExamAttempt
+      },
+      postExamAttemptHandler
+    );
+    fastify.post(
+      '/exam-environment/screenshot',
+      {
+        schema: schemas.examEnvironmentPostScreenshot
+      },
+      postScreenshotHandler
+    );
+    done();
+  };
+
+/**
+ * Wrapper for endpoints related to the exam environment desktop app.
+ *
+ * Does not require exam environment authorization token to be validated.
+ */
+export const examEnvironmentOpenRoutes: FastifyPluginCallbackTypebox = (
+  fastify,
+  _options,
+  done
+) => {
+  fastify.post(
+    '/exam-environment/token/verify',
+    {
+      schema: schemas.examEnvironmentTokenVerify
+    },
+    tokenVerifyHandler
+  );
+  done();
+};
+
+interface JwtPayload {
+  examEnvironmentAuthorizationToken: string;
+}
+
+/**
+ * Verify an authorization token has been generated for a user.
+ *
+ * Does not require any authentication.
+ *
+ * **Note**: This has no guarantees of which user the token is for. Just that one exists in the database.
+ */
+async function tokenVerifyHandler(
+  this: FastifyInstance,
+  req: UpdateReqType<typeof schemas.examEnvironmentTokenVerify>,
+  reply: FastifyReply
+) {
+  const { 'exam-environment-authorization-token': encodedToken } = req.headers;
+
+  try {
+    jwt.verify(encodedToken, JWT_SECRET);
+  } catch (e) {
+    // TODO: What to send back here? Request is valid, but token is not?
+    void reply.code(200);
+    return reply.send(
+      ERRORS.FCC_EINVAL_EXAM_ENVIRONMENT_AUTHORIZATION_TOKEN(JSON.stringify(e))
+    );
+  }
+
+  const payload = jwt.decode(encodedToken) as JwtPayload;
+
+  const examEnvironmentAuthorizationToken =
+    payload.examEnvironmentAuthorizationToken;
+
+  const token = await this.prisma.examEnvironmentAuthorizationToken.findFirst({
+    where: {
+      id: examEnvironmentAuthorizationToken
+    }
+  });
+
+  if (!token) {
+    void reply.code(200);
+    return reply.send({
+      data: 'Token does not appear to have been created.'
+    });
+  } else {
+    void reply.code(200);
+    return reply.send({
+      data: {
+        createdDate: token.createdDate
+      }
+    });
+  }
+}
+
+/**
+ * Generates an exam for the user.
+ *
+ * Requires token to be validated and TODO: live longer than the exam attempt.
+ */
+async function postExamGeneratedExamHandler(
```

Comment:
Not for this PR, but this is a huge function and it would be much easier to understand if at least some of the logic was split into a few smaller functions.